Generate platform-specific gems from a single traject.gemspec

billdueber opened the pull request:
I looked at how Oga does it, and I think it's as easy as doing what I've done here. We just need to remember to release using both MRI and JRuby.

I ran `rake build` with both MRI and JRuby and untarred the resulting gemfiles (`traject-2.0.pre-java.gem` and `traject-2.0.pre.gem`) to look at them. The metadata  (e.g., list of dependencies) all looks good. I verified that `require "traject"` under jruby does, in fact, load up the marc4j and solrj stuff (required only under JRuby in `index.rb`). 

I _think_ all we need to do is decide what the defaults should be under JRuby, set them in a `if defined? JRUBY_VERSION` block and we're home free.

jrochkind commented:
Yeah, that was my conclusion as well, I got to where you got to. 

Initially I was hoping we could make it so a rake task like `rake release` would release both platforms, to avoid user error. But that doesn't seem to be possible, as you do need to switch to jruby to do the jruby release. 

So I am a bit worried about us forgetting to do the jruby release (or vice versa) when we do a release, since you'll need to remember to do the release twice, once under jruby.  Do you think that's tolerable though, Bill?

Bah, I don't know.  Maybe I could live with the 'manual' solution instead, with one gem and the need to manually add teh marc4jreader gem.  We could have traject _try_ to require the marc4jreader when on Jruby, and if it is found use it as the default, and if it's not found print a warning message recommending it. That would lessen the 'manual' steps someone had to do. I still don't like it without bundler, as users can run into dependency version problems and wind up really needing Bundler to save them the headache. (Without bundler, and with just a separate gem with no dependency from traject gemspec, if there's a bug in marc4jreader and a later version is released -- there's no way to automatically require that, the user just has to go and make sure the 'right' version is the latest installed on their system(s)). 

I dunno, what are your current thoughts?

jrochkind commented:
oh, and I actually wouldn't bother pulling in the traject-solrj_writer.  

As far as we know, there are no general performance improvements for the solrJWriter, right? We don't actually recommend it?  If it's not the recommended-by-use for-most-use-cases solution, I don't think we need to automatically pull it in, and there are benefits to keeping MRI and Jruby on the same codebase as much as possible, when not required for performance. So I'd let people pull in SolrJWriter manually if they want it.

billdueber commented:
I'm happy to be Release master and in charge of remembering to release both. I think that's the way to go.  I'll take out solrj, set up the defaults for JRuby, add some tests, and ping you to take a look.

billdueber commented:
I've added some minor tests and changed the default JRuby reader to be Marc4J::Forgiving. I *think* there might be a problem where rubymarc outputs empty subfields? I'm gonna look into that, but for the moment I think this is a better default.

Also did a minor refactor of the default settings to more easily (if less prettily) set up with different defaults for MRI and JRuby.

jrochkind commented:
> changed the default JRuby reader to be Marc4J::Forgiving. I think there might be a problem where rubymarc outputs empty subfields? I'm gonna look into that, but for the moment I think this is a better default.

I have zero idea what you were talking about there, but that might be okay, just warning you that I got nothing from that sentance, haha. 

I'd like to review your code, but I'm pretty slammed this week. I'm also still okay with releasing a pre1 at any point, if people need it.

billdueber commented:
Heh. I can see why that made no sense. What I meant:

* I made the JRuby default the marc4j reader
* I constructed a marc-binary file by using ruby-marc to read in a marc-xml file and spit out marc21
* I tried to read in the resulting file with marc4j, but got "empty subfield" errors
* ...which leads me to believe that ruby-marc's writer outputs empty subfields in some way
* ...and also let me to make one of the default JRuby settings `'marc4j_reader.permissive' => true`

jrochkind commented:
Cool, so 

1) There might be a bug in ruby-marc writer, it happens, and ideally we'd find it and write a failing test case and file it as an issue. 

2) Was permissive the default for Marc4JReader in traject 1.x?  I think we should keep it the same. If permissive wasn't the default then, I wouldn't make it the default now -- I doubt reading files written by ruby-marc is a common use case?

billdueber commented:
I'll have to look to see if it was the default before. Permissive is also a
titch faster :-)

On Tue, Feb 10, 2015 at 4:22 PM, Jonathan Rochkind <notifications@github.com
> wrote:

> Cool, so
>
> 1) There might be a bug in ruby-marc writer, it happens, and ideally we'd
> find it and write a failing test case and file it as an issue.
>
> 2) Was permissive the default for Marc4JReader in traject 1.x? I think we
> should keep it the same. If permissive wasn't the default then, I wouldn't
> make it the default now -- I doubt reading files written by ruby-marc is a
> common use case?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/traject-project/traject/pull/83#issuecomment-73786402>
> .
>



-- 
Bill Dueber
Library Systems Programmer
University of Michigan Library
